src/bucket/BucketList.cpp

```diff
@@ -371,9 +374,151 @@ BucketList::getHash() const
+{
+    ZoneScoped;
+    for (size_t i = 0; i < mLevels.size(); ++i)
+    {
+        auto const& lev = mLevels[i];
+        std::array<std::shared_ptr<Bucket>, 2> buckets = {lev.getCurr(),
+                                                          lev.getSnap()};
+        for (auto b : buckets)
+        {
+            auto be = b->getBucketEntry(k);
+            if (be.has_value())
+            {
+                ZoneValue(static_cast<int64_t>(i));
+                return be.value().type() == DEADENTRY
+                           ? std::nullopt
+                           : std::make_optional(be.value().liveEntry());
+            }
+        }
+    }
+
+    std::string zoneTxt("Not found");
+    ZoneText(zoneTxt.c_str(), zoneTxt.size());
+    return std::nullopt;
+}
+
+// Since the input is sorted, we do a binary search at each level for the first
+// key in keys. If we find the entry, we remove the found key from keys and do
+// another binary search for the next key starting where we left off. If we
+// don't find the entry, we do not remove it from keys so that it will be
+// searched for again at a lower level.
+// The next iteration will add sharding and threading to this function
+std::vector<LedgerEntry>
+BucketList::loadKeys(std::set<LedgerKey, LedgerEntryIdCmp>& keys) const
+{
+    ZoneScoped;
+    std::vector<LedgerEntry> entries;
+    for (size_t i = 0; i < kNumLevels; ++i)
+    {
+        ZoneNamedN(__level, "blPrefetchLevel", true);
+        ZoneValue(static_cast<int64_t>(i));
+
+        int64_t numFound = 0;
+        auto const& lev = getLevel(i);
+        std::array<std::shared_ptr<Bucket>, 2> buckets = {lev.getCurr(),
+                                                          lev.getSnap()};
+        for (std::shared_ptr<Bucket> b : buckets)
+        {
+            auto currKeyIt = keys.begin();
+            auto const& index = b->getIndex();
+            auto indexIter = index.begin();
+            while (currKeyIt != keys.end() && indexIter != index.end())
+            {
+                auto offOp = index.scan(indexIter, *currKeyIt);
+                if (offOp)
+                {
+                    auto entryOp = b->getEntryAtOffset(
+                        *currKeyIt, *offOp, b->getIndex().getPageSize());
+                    if (entryOp)
+                    {
+                        if (entryOp->type() != DEADENTRY)
+                        {
+                            ++numFound;
+                            entries.push_back(entryOp->liveEntry());
+                        }
+
+                        currKeyIt = keys.erase(currKeyIt);
+                        continue;
+                    }
+                }
+
+                ++currKeyIt;
+            }
+        }
+
+        ZoneValue(numFound);
+        if (keys.empty())
+        {
+            break;
+        }
+    }
+
+    return entries;
+}
+
+std::vector<LedgerEntry>
+BucketList::loadPoolShareTrustLinesByAccountAndAsset(AccountID const& accountID,
```

Comment:
yeah this is buggy for a different reason.
See `LedgerTxnRoot::Impl::loadPoolShareTrustLinesByAccountAndAsset`:
it's a join between:
* pools that match on the asset
* trustlines for that account (this is where you need a range) that correspond to the pool IDs

This can be done by either looking up all trustlines that are pool shares first and then looking for pools (to match asset), or scan all pools first and then scan accounts (I suspect this is worst).

not sure why the assert in ledgerTxn didn't trip